pkg/eestream/decode.go

```diff
@@ -47,41 +80,86 @@ func (dr *decodedReader) Read(p []byte) (n int, err error) {
 		if dr.err != nil {
 			return 0, err
 		}
-		// we're going to kick off a bunch of goroutines. make a
-		// channel to catch those goroutine errors. importantly,
-		// the channel has a buffer size to contain all the errors
-		// even if we read none, so we can return without receiving
-		// every channel value
-		errs := make(chan readerError, len(dr.rs))
-		for i := range dr.rs {
-			go func(i int) {
-				// fill the buffer from the piece input
-				_, err := io.ReadFull(dr.rs[i], dr.inbufs[i])
-				errs <- readerError{i, err}
-			}(i)
-		}
-		// catch all the errors
-		inbufs := make(map[int][]byte, len(dr.inbufs))
+		// Run a selector on the readers' buffered channels
 		eofbufs := 0
-		for range dr.rs {
-			re := <-errs
-			if re.err == nil {
-				// add inbuf for decoding only if no error
-				inbufs[re.i] = dr.inbufs[re.i]
-			} else if re.err == io.EOF {
-				// keep track of inbufs at EOF
-				eofbufs++
+		inbufs := make(map[int][]byte, len(dr.chans))
+		// use reflect to select from the array of channels
+		cases := make([]reflect.SelectCase, len(dr.chans)+1)
+		// default case for non-blocking selection
+		cases[0] = reflect.SelectCase{
+			Dir: reflect.SelectDefault, Chan: reflect.Value{}}
+		for i, ch := range dr.chans {
+			cases[i+1] = reflect.SelectCase{
+				Dir: reflect.SelectRecv, Chan: reflect.ValueOf(ch)}
+		}
+		// iterate until the new block is received from enough channels
+		for len(cases) > 1 {
+			// non-blocking select - harvest available input buffers
+			chosen, value, ok := reflect.Select(cases)
+			if chosen == 0 {
+				// default case - no more input buffers available, check if we have enough
+				// TODO is required+1 enough to detect errors in polynomials of any degree?
+				if len(inbufs) >= dr.es.RequiredCount()+1 ||
+					len(inbufs) >= dr.es.TotalCount() {
+					// we have enough input buffers, fill the decoded output buffer
+					dr.outbuf, dr.err = dr.es.Decode(dr.outbuf, inbufs)
+					if dr.err == nil {
+						break
+					}
+					// TODO is there better way for error comparision?
+					if !strings.Contains(dr.err.Error(), "not enough shares") &&
+						!strings.Contains(dr.err.Error(), "too many errors") {
+						return 0, dr.err
+					}
+				}
+				// if enough readers are at EOF, return it
+				if eofbufs >= dr.es.RequiredCount() {
+					dr.err = io.EOF
+					return 0, dr.err
+				}
+				// blocking select - wait for more input buffers
+				chosen, value, ok = reflect.Select(cases[1:])
+				chosen++
```

Comment:
wait, the top of the for loop also has a select call

Reply:
Yes, the one in the top is non-blocking (including the default case), this one is blocking (excluding the default case at position 0).

Reply:
right, but doesn't this mean we're dropping whatever gets returned from chosen this time through? or does the non-blocking call immediately return whatever the blocking call returned?

Reply:
No, it's not dropped. It will just exit the if-block and will continue downwards as if the selection was done from the non-blocking select. Hence the `chosen++` shift.

Reply:
isn't line 156 closing a for loop and not an if block?

Reply:
oops no i'm an idiot

Reply:
okay problem exists between keyboard and chair. you're good